Move Ignition file generation to controller/common, fix Compression

cgwalters opened the pull request:
See https://github.com/coreos/butane/issues/332#issuecomment-1113670664

Basically, everywhere we make a "raw" Ignition file type, we need
to be explicitly setting the compression field.  This is prep
for updating butane and using it for some of our templates, which
will start compressing files automatically.

The problem is Ignition config merging is a bit broken with respect
to an unset compression field.  We need to make it the explicit
empty string.

Now, when I was digging into this I noticed that we'd ended up with
our API to make Ignition files in the `test/helpers`.  We shouldn't
have *core* code importing APIs from `test/`.

This moves the APIs into `controller/common`.

Port the kubelet config and container runtime config bits to this -
notice how much the code shrinks!

cgwalters commented on pkg/controller/common/helpers_test.go:
```diff
@@ -450,12 +451,35 @@ func TestRemoveIgnDuplicateFilesAndUnits(t *testing.T) {
 }
 
+// TestIgnitionMergeCompressed tests https://github.com/coreos/butane/issues/332
+func TestIgnitionMergeCompressed(t *testing.T) {
```
I verified this fails without the explicit `Compression: ""` in `NewIgnFile`.

cgwalters commented on pkg/controller/common/helpers.go:
```diff
@@ -765,6 +770,30 @@ func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []st
+		FileEmbedded1: ign3types.FileEmbedded1{
+			Contents: ign3types.Resource{
+				Source:      StrToPtr(dataurl.EncodeBytes(contents)),
+				Compression: StrToPtr(""),
```
:key: new line :arrow_up:

mtrmac commented on pkg/controller/common/helpers.go:
```diff
@@ -765,6 +770,30 @@ func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []st
+func NewIgnFileBytes(path string, contents []byte) ign3types.File {
+	return ign3types.File{
+		Node: ign3types.Node{
+			Path: path,
```
_Every single_ non-test user is setting `Overwrite: &true`; so I think it would be worth it to add a `enableOverwrite` boolean to the two helper functions.

mtrmac reviewed (commented):
(Just a quick look as someone not familiar with the problem space, going just by the diff and the problem description.)

ACK; apart from the bug fix, the consolidation of copy&pasted data is certainly a great improvement.

I wonder about the change to use `dataurl.EncodeBytes`. It does seem to me to be an improvement, or at worst immaterial, I’m just tempted to think that the people explicitly setting a MIME type and ASCII could have had a reason for doing so. Or… maybe not?

cgwalters commented:
> I wonder about the change to use dataurl.EncodeBytes. It does seem to me to be an improvement, or at worst immaterial, I’m just tempted to think that the people explicitly setting a MIME type and ASCII could have had a reason for doing so. Or… maybe not?

This is only interpreted by Ignition, which ends up calling
https://github.com/coreos/ignition/blob/646538087cdcf5e4980c663ecda4327cafaeca3e/internal/resource/url.go#L334
which as best I can tell, just parses the type out and exposes it via `ContentType()` which Ignition ignores.

cgwalters commented on pkg/controller/common/helpers.go:
```diff
@@ -45,6 +45,11 @@ import (
 )
 
+// StrToPtr returns a pointer to a string
+func StrToPtr(s string) *string {
```
I copied this from the version in `test/helpers.go`, but agree no need for it to be public.

(I have trouble writing a lot more on this subject without going off into a full-blown discourse into how Rust is so, so, so much better :wink: )

mtrmac replied:
(I think I’m in 100% violent agreement with much of that unsaid discourse.)

cgwalters commented on pkg/controller/common/helpers.go:
```diff
@@ -765,6 +770,30 @@ func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []st
+		},
+		FileEmbedded1: ign3types.FileEmbedded1{
+			Contents: ign3types.Resource{
+				Source:      StrToPtr(dataurl.EncodeBytes(contents)),
```
See below

cgwalters replied:
Well, hmm...actually it does mean that these files will start appearing as base64 inside MachineConfig, which will mean people need to base64 decode them manually if relevant.

It will also be a "spurious" change on upgrades, e.g. the machineconfig hash will be different, but the physical files will not.

I am not sure if either of these really matter though.

cgwalters commented on pkg/controller/common/helpers.go:
```diff
@@ -765,6 +770,30 @@ func CalculateConfigFileDiffs(oldIgnConfig, newIgnConfig *ign3types.Config) []st
 
+// NewIgnFile returns a simple ignition3 file from just path and file contents.
+// It also ensures the compression field is set to the empty string, which is
+// currently required for handling config merging correctly.
```
Probably doesn't matter, I am hopeful eventually we'll address this Ignition misfeature in a more systematic way anyways.

cgwalters commented:
Thanks for the thoughtful review!  Addressed all comments.  The e2e test failure I think is just a flake, cc https://github.com/openshift/machine-config-operator/pull/3039

mtrmac commented:
/lgtm

(The `data:` URL encoding change looks fine to me, but I’ll defer to actual knowledgeable maintainers.)

cgwalters commented:
Today's offhand thought: when we can hard switch to layering, the role of Ignition in the MCO will shrink...perhaps to nearly nil.  If we can move everything we need to bootstrap into the pointer config, the MCS goes away.  How we generate the content that goes into our images becomes very flexible - the stock templates don't have to be Ignition or butane or whatever necessarily, etc.
